### PR TITLE
Added script to test abi_check

### PR DESCRIPTION
**Pull request by Zaya-dyno**
## Description
This new script will test abi_check.
Fixes https://github.com/Mbed-TLS/mbedtls/issues/5409

## Todos
- [ ] Changes to all.sh will be made so that script can be called from there.
- [ ] Backported

## Steps to test or reproduce
run script from root of project

### CODE REVIEW

File: tests/scripts/check_abi_check.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import re
+
+SCRIPT_PATH = "scripts/abi_check.py"
+
+def run_abi_check(old_tag, new_tag):
+    return subprocess.check_output([SCRIPT_PATH, "-o", old_tag, "-n", new_tag],
+                                   stderr=subprocess.STDOUT).decode("utf-8")
+
+def pattern_create(lines):
+    return "^["+ "".join(["("+line+")" for line in lines]) +"]$"
```
Comment (gilles-peskine-arm) [inline tests/scripts/check_abi_check.py:35]: This doesn't do what you think. `[…]` is a _character_ set, so you're building a regex that matches a single character. Alternation (matching one of multiple patterns) is `…|…`.

File: tests/scripts/check_abi_check.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    def test_abi_same(self):
+        ''' This test checks no false positive
+        '''
+        possible_lines = ["Checking.*", "No compatibility issues.*", "PASS.*", "Info.*"]
+        pattern = pattern_create(possible_lines)
+        out = run_abi_check("mbedtls-3.2.0","mbedtls-3.2.1")
+        self.assertIsNone(re.fullmatch(pattern, out, re.MULTILINE))
+
+    def test_abi_different(self):
```
Comment (gilles-peskine-arm) [inline tests/scripts/check_abi_check.py:54]: `abi_check` does several things, currently 4: API differences, ABI differences, generated file differences, storage format test differences. We should have a negative test for each of them: e.g. there should at least one test case with (compatible ABI, incompatible ABI, compatible generated files, compatible storage format tests), and one with (compatible API, compatible ABI, compatible generated files, incompatible storage format tests), and one with (compatible API, compatible ABI, incompatible generated files, compatible storage format tests) (I'm not sure if incompatible API, compatible ABI is possible with abi_check's definition of API/ABI compatibility).

The task description for #5409 specifically requests negative testing for API/ABI compatibility, and there's a separate task #5410 for generated test cases. So for #5409 it's enough to focus on API/ABI differences, but the test has to fail if `abi_check.py` only reports incompatibilities in generated tests.

File: tests/scripts/check_abi_check.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+def run_main():
+    try:
+        TestAbiCheck.check_repo_path()
+        unittest.main()
+    except Exception: #pylint: disable=broad-except
+        traceback.print_exc()
+        sys.exit(2)
+
+if __name__ == "__main__":
+    run_main()
```
Comment (gilles-peskine-arm) [inline tests/scripts/check_abi_check.py:75]: We need to have a component of `all.sh` that runs this, so that it's run on the CI.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

**Comment by gilles-peskine-arm** [comment]
Ignore the CI statuses for PR-5986, due to an accidental incorrect force-push.

File: tests/scripts/check_abi_check.py (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/scripts/check_abi_check.py:54]: `_read_storage_tests` records test cases if they either come from an automatically generated file or the function name contains `read`. So for #5410, there should be a test case that removes a `read` test from a non-automatically-generated file, and a test case that removes a non-`read` test from an automatically generated file.

c4b814a9c2c192f8f52908645720ced9cb817447 is an ABI break and not an API break. All enum identifiers still exist afterwards, so source programs will still work, so it's a compatible API. But the values of some enum identifiers change, so some compiled programs won't work anymore, so it's an ABI break.

**Comment by daverodgman** [comment]
Closing since Zaya is unlikely to progress this further.

**Pull request closed without merging**